File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-4.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n=len(prices)
        if len(prices)<=1 :
            return 0
        res = 0
        d={}
        def dfs(i):
            if i in d :
                return d[i]
            if i >= n-1 :
                return 0
            d[i]=0
            for j in range(i+1,n):
                d[i]=max(d[i],(prices[j]-prices[i])+dfs(j+2))
            return d[i]
        for i in range(n-1):
            res = max(res,dfs(i))
        return res
        

File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-4.cpp
class Solution {
public: 
    unordered_map<char,string> D = {
        {'2',"abc"},
        {'3',"def"},
        {'4',"ghi"},
        {'5',"jkl"},
        {'6',"mno"},
        {'7',"pqrs"},
        {'8',"tuv"},
        {'9',"wxyz"}
    };
    vector<string> res ={};
    string cur;
    void dfs(string digits,int i,int k ){
        if(k== D[digits[i]].size()) return ;
        cur+= D[digits[i]].at(k);
        if(i==digits.size()-1){
            res.push_back(cur);
            cur.pop_back();
            dfs(digits,i,k+1);
        }
        else{
            dfs(digits,i+1,0);
            cur.pop_back();
            dfs(digits,i,k+1);
        }
    }
    vector<string> letterCombinations(string digits) {
        vector<string> x ={};
        if( digits.size() ==0) return x ;
        dfs(digits,0,0);
        return res;
    }
};
